File: tests/support/proving_profile.rs
use hugr_lean::command::{CommandRecognition, InvocationIdentity};
use hugr_lean::preservation::{
    ByteSpan, LeanWriter, PreservationContract, RenderedOutput, Signal, SignalId,
};
use hugr_lean::profile::{
    AnalysisBundle, Profile, ProfileAnalysis, ProfileContext, ProfileError, ProfileMatch,
    ProfileRequirements, RouteContext,
};

pub const PROVING_PROFILE_ID: &str = "proving-profile";
const PROOF_SIGNAL_ID: SignalId = SignalId::new("proof-signal");

pub struct ProvingProfile;

struct ProvingAnalysis {
    signal: Signal,
}

impl Profile for ProvingProfile {
    fn id(&self) -> &'static str {
        PROVING_PROFILE_ID
    }

    fn requirements(&self) -> ProfileRequirements {
        ProfileRequirements::COMPLETE_EXITED
    }

    fn recognize(&self, identity: &InvocationIdentity) -> ProfileMatch {
        match identity {
            InvocationIdentity::Shell(CommandRecognition::Direct(command))
                if command.program == "hugr-lean-prove" =>
            {
                ProfileMatch::Match
            }
            _ => ProfileMatch::NoMatch,
        }
    }

    fn shape_guard(&self, context: &RouteContext<'_>) -> ProfileMatch {
        if proof_span(&context.observation.output).is_some() {
            ProfileMatch::Match
        } else {
            ProfileMatch::NoMatch
        }
    }

    fn analyze(&self, context: &ProfileContext<'_>) -> Result<AnalysisBundle, ProfileError> {
        let span = proof_span(context.safe_baseline).ok_or_else(ProfileError::analyze)?;
        let signal = context
            .verbatim_signal(PROOF_SIGNAL_ID, span)
            .map_err(|_| ProfileError::analyze())?;

        Ok(AnalysisBundle::new(
            Box::new(ProvingAnalysis { signal }),
            PreservationContract::require(PROOF_SIGNAL_ID),
        ))
    }

    fn render(
        &self,
        analysis: &dyn ProfileAnalysis,
        writer: &mut LeanWriter,
    ) -> Result<(), ProfileError> {
        let analysis = analysis
            .as_any()
            .downcast_ref::<ProvingAnalysis>()
            .ok_or_else(ProfileError::render)?;

        writer.signal(&analysis.signal);
        writer.newline();
        Ok(())
    }

    fn validate(
        &self,
        analysis: &dyn ProfileAnalysis,
        rendered: &RenderedOutput,
    ) -> Result<(), ProfileError> {
        let analysis = analysis
            .as_any()
            .downcast_ref::<ProvingAnalysis>()
            .ok_or_else(ProfileError::validate)?;

        let expected = format!("{}\n", analysis.signal.canonical_text());
        if rendered.text() != expected || !rendered.emitted_signal_ids().contains(&PROOF_SIGNAL_ID)
        {
            return Err(ProfileError::validate());
        }

        Ok(())
    }
}

fn proof_span(input: &str) -> Option<ByteSpan> {
    let mut offset = 0;

    for line in input.split_inclusive('\n') {
        let content = line.strip_suffix('\n').unwrap_or(line);
        if content.starts_with("PROOF ") {
            return Some(ByteSpan::new(offset, offset + content.len()));
        }
        offset += line.len();
    }

    if offset < input.len() {
        let content = &input[offset..];
        if content.starts_with("PROOF ") {
            return Some(ByteSpan::new(offset, input.len()));
        }
    }

    None
}

File: tests/support/proving-profile.ts
import {
  LeanWriter,
  PreservationContract,
  signalId,
  type Signal,
} from "../../src/preservation.js";
import {
  COMPLETE_EXITED,
  ProfileContext,
  type AnalysisBundle,
  type Profile,
  type ProfileDescriptor,
  type RouteContext,
} from "../../src/profile.js";

export const PROVING_PROFILE_ID = "proving-profile";
export const PROOF_SIGNAL_NAME = "proof-signal";
const proofSignalId = signalId(PROOF_SIGNAL_NAME);

interface ProvingAnalysis {
  signal: Signal;
}

export class ProvingProfile implements Profile {
  descriptor(): ProfileDescriptor {
    return {
      id: PROVING_PROFILE_ID,
      family: "test",
      fixture_family: "proving",
      boundary_assumption: "native_text",
    };
  }

  requirements() {
    return COMPLETE_EXITED;
  }

  recognize(identity: Parameters<Profile["recognize"]>[0]) {
    return identity.kind === "shell" &&
      identity.recognition.kind === "direct" &&
      identity.recognition.identity.program === "hugr-lean-prove"
      ? ("match" as const)
      : ("no_match" as const);
  }

  shapeGuard(context: RouteContext) {
    return proofLineSpan(context.safe_baseline) === null
      ? ("no_match" as const)
      : ("match" as const);
  }

  analyze(context: ProfileContext): AnalysisBundle {
    const span = proofLineSpan(context.safe_baseline);
    if (span === null) {
      throw new Error("proof line missing");
    }

    const signal = context.verbatimSignal(proofSignalId, span);
    return {
      data: { signal } satisfies ProvingAnalysis,
      preservation: PreservationContract.require(proofSignalId),
    };
  }

  render(analysis: unknown, writer: LeanWriter): void {
    const value = analysis as ProvingAnalysis;
    writer.signalLine(value.signal);
  }

  validate(analysis: unknown, rendered: ReturnType<LeanWriter["finish"]>): void {
    const value = analysis as ProvingAnalysis;
    if (
      rendered.text !== value.signal.canonical_text + "\n" ||
      !rendered.emitted_signal_ids.has(proofSignalId)
    ) {
      throw new Error("proving profile validation failed");
    }
  }
}

function proofLineSpan(input: string): { start_byte: number; end_byte: number } | null {
  let codeUnitOffset = 0;
  let byteOffset = 0;

  while (codeUnitOffset < input.length) {
    const newline = input.indexOf("\n", codeUnitOffset);
    const chunkEnd = newline < 0 ? input.length : newline + 1;
    const chunk = input.slice(codeUnitOffset, chunkEnd);
    const hasLf = chunk.endsWith("\n");
    const withoutLf = hasLf ? chunk.slice(0, -1) : chunk;
    const content =
      hasLf && withoutLf.endsWith("\r")
        ? withoutLf.slice(0, -1)
        : withoutLf;

    if (content.startsWith("PROOF ")) {
      return {
        start_byte: byteOffset,
        end_byte: byteOffset + Buffer.byteLength(content, "utf8"),
      };
    }

    codeUnitOffset = chunkEnd;
    byteOffset += Buffer.byteLength(chunk, "utf8");
  }

  return null;
}
